workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentSenderGenerator.tt

```diff
@@ -0,0 +1,85 @@
+<#@ template language="C#" #>
+<#@ output extension=".cs" #>
+<#
+    var componentDetails = GetComponentDetails();
+    var generatedHeader = CommonGeneratorUtils.GetGeneratedHeader();
+    var componentNamespace = qualifiedNamespace + "." + componentDetails.ComponentName;
+    var profilingStart = $"Profiler.BeginSample(\"{componentDetails.ComponentName}\");";
+    var profilingEnd = "Profiler.EndSample();";
+#>
+<#= generatedHeader #>
+
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Profiling;
+using Unity.Mathematics;
+using Unity.Entities;
+using Unity.Collections;
+using Improbable.Gdk.Core;
+using Improbable.Gdk.Core.CodegenAdapters;
+
+namespace <#= qualifiedNamespace #>
+{
+    public partial class <#= componentDetails.ComponentName #>
+    {
+        internal class ComponentReplicator : IComponentReplicationHandler
+        {
+            public uint ComponentId => <#= componentDetails.ComponentId #>;
+
+            public EntityArchetypeQuery ComponentUpdateQuery => new EntityArchetypeQuery
+            {
+                All = new[]
+                {
+                    ComponentType.Create<<#= componentNamespace #>.Component>(),
+                    ComponentType.Create<<#= componentNamespace #>.ComponentAuthority>(),
+                    ComponentType.ReadOnly<SpatialEntityId>()
+                },
+                Any = Array.Empty<ComponentType>(),
+                None = Array.Empty<ComponentType>(),
+            };
+
+            public void SendUpdates(
+                ComponentGroup replicationGroup,
+                ComponentSystemBase system,
+                EntityManager entityManager,
+                ComponentUpdateSystem componentUpdateSystem)
+            {
+                <#= profilingStart #>
+
+                var chunkArray = replicationGroup.CreateArchetypeChunkArray(Allocator.TempJob);
+                var spatialOSEntityType = system.GetArchetypeChunkComponentType<SpatialEntityId>(true);
+                var componentType = system.GetArchetypeChunkComponentType<<#= componentNamespace #>.Component>();
+
+                var authorityType = system.GetArchetypeChunkSharedComponentType<ComponentAuthority>();
+
+                foreach (var chunk in chunkArray)
+                {
+                    var entityIdArray = chunk.GetNativeArray(spatialOSEntityType);
+                    var componentArray = chunk.GetNativeArray(componentType);
+
+                    var authorityIndex = chunk.GetSharedComponentIndex(authorityType);
+
+                    if (!entityManager.GetSharedComponentData<ComponentAuthority>(authorityIndex).HasAuthority)
+                    {
+                        continue;
```

Comment:
should we print a warning here? or do we expect this to happen when e.g. authority just changed?

Reply:
Thing the above should say - this check is applied per chunk, not per entity.